File: apps/server/internal/store/evidence_validation.go
package store

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"math"
	"reflect"
	"regexp"
	"strings"
)

//go:embed evidence-schema.json
var evidenceSchemaJSON []byte

type wireSchema struct {
	Ref        string                `json:"$ref"`
	Type       string                `json:"type"`
	Const      json.RawMessage       `json:"const"`
	AnyOf      []wireSchema          `json:"anyOf"`
	Properties map[string]wireSchema `json:"properties"`
	Required   []string              `json:"required"`
	Additional json.RawMessage       `json:"additionalProperties"`
	Items      *wireSchema           `json:"items"`
	Minimum    *float64              `json:"minimum"`
	Maximum    *float64              `json:"maximum"`
	MinLength  int                   `json:"minLength"`
	Pattern    string                `json:"pattern"`
}

var evidenceSchemas = func() map[string]wireSchema {
	var schemas map[string]wireSchema
	if err := json.Unmarshal(evidenceSchemaJSON, &schemas); err != nil {
		panic(err)
	}
	return schemas
}()

func ValidateEvidenceJSON(name string, raw []byte) error {
	schema, ok := evidenceSchemas[name]
	if !ok {
		return fmt.Errorf("unknown evidence model %s", name)
	}
	var value any
	if err := json.Unmarshal(raw, &value); err != nil {
		return err
	}
	if err := validateWire(schema, value, name, 0); err != nil {
		return err
	}
	if name == "ContractContent" || name == "IssueContract" {
		var content ContractContent
		if err := json.Unmarshal(raw, &content); err != nil {
			return err
		}
		confirmed := false
		if name == "IssueContract" {
			var contract IssueContract
			if err := json.Unmarshal(raw, &contract); err != nil {
				return err
			}
			confirmed = contract.Status == "confirmed"
			if confirmed && (contract.Confirmation == nil || contract.Confirmation.ContentDigest != contract.ContentDigest || !IsHumanActor(contract.Confirmation.Actor)) {
				return fmt.Errorf("confirmed contract requires exact human confirmation")
			}
			if (contract.Status == "draft" || contract.Status == "discarded") && contract.Confirmation != nil {
				return fmt.Errorf("unconfirmed draft cannot carry a confirmation")
			}
		}
		if err := ValidateContractContent(content, confirmed); err != nil {
			return err
		}
	}
	if name == "Verification" {
		var v Verification
		if err := json.Unmarshal(raw, &v); err != nil {
			return err
		}
		if (v.Status == "completed") != (v.Result != nil) {
			return fmt.Errorf("only completed verification has a result")
		}
		if v.Mode == "deterministic" && v.Executor.Kind != "program" {
			return fmt.Errorf("deterministic verification requires program")
		}
		if v.Mode == "agent" && v.Executor.Kind != "agent" {
			return fmt.Errorf("agent verification requires isolated agent")
		}
		if v.Status == "failed" && v.Error == nil {
			return fmt.Errorf("failed verification requires technical error")
		}
	}
	return validateNestedSelectors(value)
}

func validateWire(s wireSchema, value any, path string, depth int) error {
	bad := func(why string) error { return fmt.Errorf("%s: %s", path, why) }
	if depth > 80 {
		return bad("nesting limit exceeded")
	}
	if s.Ref != "" {
		return validateWire(evidenceSchemas[s.Ref], value, path, depth+1)
	}
	if len(s.AnyOf) > 0 {
		for _, variant := range s.AnyOf {
			if validateWire(variant, value, path, depth+1) == nil {
				return nil
			}
		}
		return bad("invalid union variant")
	}
	if s.Const != nil {
		var expected any
		_ = json.Unmarshal(s.Const, &expected)
		if !reflect.DeepEqual(value, expected) {
			return bad("unexpected literal")
		}
		return nil
	}
	switch s.Type {
	case "object":
		obj, ok := value.(map[string]any)
		if !ok {
			return bad("expected object")
		}
		for _, key := range s.Required {
			if _, ok := obj[key]; !ok {
				return bad("missing " + key)
			}
		}
		for key, item := range obj {
			property, known := s.Properties[key]
			if !known && len(s.Additional) > 0 && s.Additional[0] == '{' {
				_ = json.Unmarshal(s.Additional, &property)
				known = true
			}
			if !known {
				return bad("unknown field " + key)
			}
			if err := validateWire(property, item, path+"."+key, depth+1); err != nil {
				return err
			}
		}
	case "array":
		items, ok := value.([]any)
		if !ok {
			return bad("expected array")
		}
		for i, item := range items {
			if err := validateWire(*s.Items, item, fmt.Sprintf("%s[%d]", path, i), depth+1); err != nil {
				return err
			}
		}
	case "string":
		str, ok := value.(string)
		if !ok {
			return bad("expected string")
		}
		if s.MinLength > 0 && strings.TrimSpace(str) == "" {
			return bad("empty string")
		}
		if s.Pattern != "" && !regexp.MustCompile(s.Pattern).MatchString(str) {
			return bad("invalid format")
		}
	case "boolean":
		if _, ok := value.(bool); !ok {
			return bad("expected boolean")
		}
	case "number", "integer":
		n, ok := value.(float64)
		if !ok || math.IsInf(n, 0) || math.IsNaN(n) || (s.Type == "integer" && (n != math.Trunc(n) || math.Abs(n) > 9007199254740991)) || (s.Minimum != nil && n < *s.Minimum) || (s.Maximum != nil && n > *s.Maximum) {
			return bad("invalid number")
		}
	default:
		return bad("unsupported schema")
	}
	return nil
}

func validateNestedSelectors(value any) error {
	switch v := value.(type) {
	case []any:
		for _, child := range v {
			if err := validateNestedSelectors(child); err != nil {
				return err
			}
		}
	case map[string]any:
		kind, _ := v["kind"].(string)
		num := func(k string) float64 { n, _ := v[k].(float64); return n }
		if kind == "text_lines" && num("end") < num("start") {
			return fmt.Errorf("reversed line range")
		}
		if kind == "time_range" && num("endMs") <= num("startMs") {
			return fmt.Errorf("empty time range")
		}
		if kind == "image_region" && (num("width") <= 0 || num("height") <= 0 || num("x")+num("width") > 1 || num("y")+num("height") > 1) {
			return fmt.Errorf("image region out of bounds")
		}
		if kind == "json_pointer" {
			p, _ := v["pointer"].(string)
			if !regexp.MustCompile(`^(?:/(?:[^~]|~[01])*)*$`).MatchString(p) {
				return fmt.Errorf("invalid JSON pointer")
			}
		}
		for _, child := range v {
			if err := validateNestedSelectors(child); err != nil {
				return err
			}
		}
	}
	return nil
}

// Go unions preserve the discriminator's required empty arrays/maps on marshal.
func marshalEvidenceUnion(name, kind string, raw []byte) ([]byte, error) {
	var object map[string]any
	if err := json.Unmarshal(raw, &object); err != nil {
		return nil, err
	}
	for _, branch := range evidenceSchemas[name].AnyOf {
		if string(branch.Properties["kind"].Const) != fmt.Sprintf("%q", kind) {
			continue
		}
		for _, key := range branch.Required {
			if object[key] != nil {
				continue
			}
			switch branch.Properties[key].Type {
			case "array":
				object[key] = []any{}
			case "object":
				object[key] = map[string]any{}
			}
		}
	}
	return json.Marshal(object)
}

func IsHumanActor(actor ActorRef) bool {
	return actor.ID != "" && (actor.Kind == "user" || actor.Kind == "local_owner")
}

func ValidateContractContent(c ContractContent, confirming bool) error {
	if c.Criteria == nil || c.Goal.Media == nil || c.InScope == nil || c.OutOfScope == nil || c.Constraints == nil {
		return fmt.Errorf("contract arrays must not be null")
	}
	ids := map[string]bool{}
	required := 0
	for _, criterion := range c.Criteria {
		if ids[criterion.ID] {
			return fmt.Errorf("duplicate criterion %s", criterion.ID)
		}
		ids[criterion.ID] = true
		if criterion.Required {
			required++
		}
		if strings.TrimSpace(criterion.Statement) == "" || strings.TrimSpace(criterion.Rubric.Text) == "" {
			return fmt.Errorf("observable statement and rubric required")
		}
		if criterion.ProofKind == "other" && (criterion.ProofKindLabel == nil || strings.TrimSpace(*criterion.ProofKindLabel) == "") {
			return fmt.Errorf("proofKindLabel required")
		}
		if criterion.EvaluationMode == "agent" && criterion.Checker != nil {
			return fmt.Errorf("agent criterion cannot have checker")
		}
		// A deterministic criterion without its checker can never be verified;
		// confirming one would deadlock acceptance.
		if confirming && criterion.EvaluationMode == "deterministic" && criterion.Checker == nil {
			return fmt.Errorf("deterministic criterion requires a runnable checker")
		}
		if len(criterion.EvidenceRequirements) == 0 {
			return fmt.Errorf("evidence requirement required")
		}
		requirements := map[string]bool{}
		for _, r := range criterion.EvidenceRequirements {
			if requirements[r.ID] || len(r.AcceptedCarriers) == 0 || r.MinimumCount < 1 {
				return fmt.Errorf("invalid or duplicate evidence requirement")
			}
			requirements[r.ID] = true
			if criterion.EvaluationMode == "deterministic" && r.BindingPolicy != "system_observed" {
				return fmt.Errorf("deterministic evidence must be system observed")
			}
		}
		if checker := criterion.Checker; checker != nil && (checker.Configuration.Kind == "command" || checker.Configuration.Kind == "project_command") {
			config := checker.Configuration
			paths := []*string{config.CwdRelativePath}
			if config.Kind == "command" {
				paths = append(paths, config.Entrypoint)
			}
			for _, path := range paths {
				if path == nil || *path == "" || strings.HasPrefix(*path, "/") || strings.Contains(*path, `\`) {
					return fmt.Errorf("unsafe checker path")
				}
				for _, part := range strings.Split(*path, "/") {
					if part == ".." {
						return fmt.Errorf("unsafe checker path")
					}
				}
			}
			if len(config.ExpectedExitCodes) == 0 {
				return fmt.Errorf("expected exit codes required")
			}
			if config.Executable == nil || strings.TrimSpace(*config.Executable) == "" {
				return fmt.Errorf("checker executable required")
			}
			// The project's own command carries no checker bundle, fixtures or
			// secret bindings; anything else would never be run as declared.
			if config.Kind == "project_command" && (config.CheckerBundleMaterialID != nil ||
				len(config.FixtureMaterialIDs) > 0 || len(config.SecretBindings) > 0 ||
				config.ResultFormat != nil || config.MinimumAssertions != nil) {
				return fmt.Errorf("project command takes only executable, args, cwd, environment and exit codes")
			}
		}
	}
	if confirming && (strings.TrimSpace(c.Goal.Text) == "" || required == 0) {
		return fmt.Errorf("confirmation requires goal and at least one required criterion")
	}
	raw, _ := json.Marshal(c)
	var value any
	_ = json.Unmarshal(raw, &value)
	return validateNestedSelectors(value)
}

File: packages/protocol/src/evidence-validation.ts
import definitions from "./evidence-schema.json" with { type: "json" };
import type {
  ContractContent,
  IssueContract,
  MaterialSelector,
  Verification,
} from "./evidence.js";

type Schema = {
  $ref?: string;
  type?: string;
  const?: unknown;
  anyOf?: Schema[];
  properties?: Record<string, Schema>;
  required?: string[];
  additionalProperties?: boolean | Schema;
  items?: Schema;
  minimum?: number;
  maximum?: number;
  minLength?: number;
  pattern?: string;
};
const schemas = definitions as unknown as Record<string, Schema>;

/** Only the selected model and its referenced definitions, for structured prompts. */
export function evidenceModelSchema(
  name: keyof typeof definitions,
): Record<string, unknown> {
  const selected: Record<string, unknown> = {};
  const visit = (value: unknown): void => {
    if (!value || typeof value !== "object") return;
    if (Array.isArray(value)) {
      value.forEach(visit);
      return;
    }
    const object = value as Record<string, unknown>;
    if (typeof object.$ref === "string" && !selected[object.$ref]) {
      selected[object.$ref] = schemas[object.$ref];
      visit(schemas[object.$ref]);
    }
    Object.values(object).forEach(visit);
  };
  selected[name] = schemas[name];
  visit(schemas[name]);
  return selected;
}

/**
 * The same model as a self-contained standard JSON Schema, for providers that
 * constrain generation to a schema instead of trusting prose instructions.
 */
export function evidenceJSONSchema(
  name: keyof typeof definitions,
  omitProperties: string[] = [],
): Record<string, unknown> {
  const rewrite = (value: unknown): unknown => {
    if (Array.isArray(value)) return value.map(rewrite);
    if (!value || typeof value !== "object") return value;
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>).map(([key, entry]) =>
        key === "$ref" && typeof entry === "string"
          ? [key, `#/$defs/${entry}`]
          : [key, rewrite(entry)],
      ),
    );
  };
  const bundle = rewrite(evidenceModelSchema(name)) as Record<
    string,
    Record<string, unknown>
  >;
  const root = { ...bundle[name] };
  if (omitProperties.length) {
    root.properties = Object.fromEntries(
      Object.entries((root.properties ?? {}) as Record<string, unknown>).filter(
        ([key]) => !omitProperties.includes(key),
      ),
    );
    root.required = ((root.required ?? []) as string[]).filter(
      (key) => !omitProperties.includes(key),
    );
  }
  return { ...root, $defs: bundle };
}

/** Strict wire validation: unknown fields and null arrays are errors. */
export function validateEvidenceModel(
  name: keyof typeof definitions,
  value: unknown,
): string[] {
  const errors = check(schemas[name]!, value, name, 0);
  if (errors.length) return errors;
  errors.push(...nestedSelectors(value));
  if (name === "ContractContent")
    errors.push(...validateContractContent(value as ContractContent));
  if (name === "IssueContract") {
    const contract = value as IssueContract;
    errors.push(
      ...validateContractContent(contract, contract.status === "confirmed"),
    );
    if (
      contract.status === "confirmed" &&
      (!contract.confirmation ||
        contract.confirmation.contentDigest !== contract.contentDigest ||
        !["user", "local_owner"].includes(contract.confirmation.actor.kind))
    )
      errors.push("Confirmed contract requires exact human confirmation");
    if (
      ["draft", "discarded"].includes(contract.status) &&
      contract.confirmation
    )
      errors.push("Unconfirmed draft cannot carry a confirmation");
  }
  if (name === "MaterialSelector")
    errors.push(...validateSelector(value as MaterialSelector));
  if (name === "Verification") {
    const v = value as Verification;
    if ((v.status === "completed") !== !!v.result)
      errors.push("Only completed verification has a result");
    if (v.mode === "deterministic" && v.executor.kind !== "program")
      errors.push("Deterministic verification requires program");
    if (v.mode === "agent" && v.executor.kind !== "agent")
      errors.push("Agent verification requires isolated agent");
    if (v.status === "failed" && !v.error)
      errors.push("Failed verification requires technical error");
  }
  return errors;
}
function nestedSelectors(value: unknown): string[] {
  if (Array.isArray(value)) return value.flatMap(nestedSelectors);
  if (!value || typeof value !== "object") return [];
  const object = value as Record<string, unknown>;
  const errors = [
    "text_lines",
    "page",
    "time_range",
    "image_region",
    "json_pointer",
  ].includes(String(object.kind))
    ? validateSelector(value as MaterialSelector)
    : [];
  return [...errors, ...Object.values(object).flatMap(nestedSelectors)];
}

function check(s: Schema, v: unknown, path: string, depth: number): string[] {
  if (depth > 80) return [`${path}: nesting limit exceeded`];
  if (s.$ref) return check(schemas[s.$ref]!, v, path, depth + 1);
  if (s.anyOf)
    return s.anyOf.some((b) => check(b, v, path, depth + 1).length === 0)
      ? []
      : [`${path}: invalid union variant`];
  if ("const" in s)
    return v === s.const
      ? []
      : [`${path}: expected ${JSON.stringify(s.const)}`];
  if (s.type === "object") {
    if (!v || typeof v !== "object" || Array.isArray(v))
      return [`${path}: expected object`];
    const object = v as Record<string, unknown>;
    const errors = (s.required ?? [])
      .filter((k) => !Object.hasOwn(object, k))
      .map((k) => `${path}.${k}: required`);
    for (const [k, value] of Object.entries(object)) {
      if (value === undefined && s.properties?.[k] && !s.required?.includes(k))
        continue;
      const field =
        s.properties?.[k] ??
        (typeof s.additionalProperties === "object"
          ? s.additionalProperties
          : undefined);
      if (!field) errors.push(`${path}.${k}: unknown field`);
      else errors.push(...check(field, value, `${path}.${k}`, depth + 1));
    }
    return errors;
  }
  if (s.type === "array")
    return Array.isArray(v)
      ? v.flatMap((x, i) => check(s.items!, x, `${path}[${i}]`, depth + 1))
      : [`${path}: expected array`];
  if (s.type === "string") {
    if (typeof v !== "string") return [`${path}: expected string`];
    if (s.minLength && !v.trim()) return [`${path}: empty string`];
    if (s.pattern && !new RegExp(s.pattern).test(v))
      return [`${path}: invalid format`];
  } else if (s.type === "boolean") {
    if (typeof v !== "boolean") return [`${path}: expected boolean`];
  } else if (
    typeof v !== "number" ||
    !Number.isFinite(v) ||
    (s.type === "integer" && !Number.isSafeInteger(v)) ||
    (s.minimum !== undefined && v < s.minimum) ||
    (s.maximum !== undefined && v > s.maximum)
  ) {
    return [`${path}: invalid number`];
  }
  return [];
}

export function validateSelector(s: MaterialSelector): string[] {
  if (s.kind === "text_lines" && s.end < s.start)
    return ["Reversed line range"];
  if (s.kind === "time_range" && s.endMs <= s.startMs)
    return ["Empty or reversed time range"];
  if (
    s.kind === "image_region" &&
    (s.width <= 0 || s.height <= 0 || s.x + s.width > 1 || s.y + s.height > 1)
  )
    return ["Image region out of bounds"];
  if (s.kind === "json_pointer" && !/^(?:\/(?:[^~]|~[01])*)*$/.test(s.pointer))
    return ["Invalid JSON Pointer"];
  return [];
}

export function validateContractContent(
  content: ContractContent,
  confirming = false,
): string[] {
  const errors: string[] = [];
  const ids = new Set<string>();
  if (
    confirming &&
    (!content.goal.text.trim() || !content.criteria.some((c) => c.required))
  )
    errors.push(
      "Confirmation requires a goal and at least one required criterion",
    );
  for (const c of content.criteria) {
    if (ids.has(c.id)) errors.push(`Duplicate criterion ${c.id}`);
    ids.add(c.id);
    if (!c.statement.trim() || !c.rubric.text.trim())
      errors.push(`${c.id}: observable statement and rubric required`);
    if (c.proofKind === "other" && !c.proofKindLabel?.trim())
      errors.push(`${c.id}: proofKindLabel required`);
    if (!c.evidenceRequirements.length)
      errors.push(`${c.id}: evidence requirement required`);
    const requirements = new Set<string>();
    for (const r of c.evidenceRequirements) {
      if (requirements.has(r.id))
        errors.push(`${c.id}: duplicate requirement ${r.id}`);
      requirements.add(r.id);
      if (!r.acceptedCarriers.length)
        errors.push(`${c.id}: accepted carriers required`);
      if (
        c.evaluationMode === "deterministic" &&
        r.bindingPolicy !== "system_observed"
      )
        errors.push(`${c.id}: deterministic evidence must be system observed`);
    }
    if (c.evaluationMode === "agent" && c.checker)
      errors.push(`${c.id}: agent criterion cannot have checker`);
    if (c.checker?.configuration.kind === "command") {
      const config = c.checker.configuration;
      for (const path of [config.cwdRelativePath, config.entrypoint]) {
        if (
          !path ||
          path.startsWith("/") ||
          path.includes("\\") ||
          path.split("/").includes("..")
        )
          errors.push(`${c.id}: unsafe checker path`);
      }
      if (!config.expectedExitCodes.length)
        errors.push(`${c.id}: expected exit codes required`);
    }
    for (const media of c.rubric.media)
      if (media.selector) errors.push(...validateSelector(media.selector));
  }
  for (const media of content.goal.media)
    if (media.selector) errors.push(...validateSelector(media.selector));
  return errors;
}
